compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -201,29 +246,29 @@ abstract class GenJSCode extends plugins.PluginComponent
         }
         val allClassDefs = collectClassDefs(cunit.body)
 
-        /* First gen and record lambdas for js.FunctionN and js.ThisFunctionN.
-         * Since they are SAMs, there cannot be dependencies within this set,
-         * and hence we are sure we can record them before they are used,
-         * which is critical for these.
+        /* There are three types of anonymous classes we want to generate
+         * only when we need them so we can inline them at construction site:
+         *
+         * - lambdas for js.FunctionN and js.ThisFunctionN (SAMs).
+         * - lambdas for scala.FunctionN. Note that these may fail.
+         * - anonymous Scala.js defined classes. We'll inline the constructor.
+         *
+         * Note that the first two are optimizations (we'd rather have a cheap
+         * JS lambda, than a fully fleged scala class).
```

Comment:
The first one is a requirement, too. We cannot generate an actual class extending `js.(This)FunctionN`. We must recover them into lambdas.